File: frontend/src/utils/roleUtils.js
// Role-based access control utilities
// Centralized role management for the Carbon Credit Verification System

// Define all available roles as constants
export const ROLES = {
  // Admin roles (legacy support)
  ADMINISTRATOR: 'Administrator', // Standard admin role
  ADMIN: 'admin', // Legacy lowercase admin role
  
  // User roles
  PROJECT_DEVELOPER: 'Project Developer',
  VERIFIER: 'Verifier',
  SCIENTIST: 'Scientist',
  RESEARCHER: 'Researcher',
  INVESTOR: 'Investor',
  BROKER: 'Broker',
  REGULATOR: 'Regulator',
  MONITOR: 'Monitor',
  AUDITOR: 'Auditor'
};

// Define role hierarchies and permissions
export const ROLE_PERMISSIONS = {
  // Administrative roles - highest permissions
  ADMIN_ROLES: [ROLES.ADMINISTRATOR, ROLES.ADMIN],
  
  // Verification and quality assurance
  VERIFICATION_ROLES: [ROLES.VERIFIER, ROLES.AUDITOR],
  
  // Scientific and research roles
  SCIENTIFIC_ROLES: [ROLES.SCIENTIST, ROLES.RESEARCHER],
  
  // Business and finance roles
  BUSINESS_ROLES: [ROLES.INVESTOR, ROLES.BROKER],
  
  // Regulatory and compliance roles
  REGULATORY_ROLES: [ROLES.REGULATOR, ROLES.MONITOR],
  
  // Development roles
  DEVELOPMENT_ROLES: [ROLES.PROJECT_DEVELOPER],
  
  // All roles that can access basic features
  ALL_AUTHENTICATED_ROLES: Object.values(ROLES)
};

// Define feature access groups
export const FEATURE_ACCESS = {
  // Core features accessible to most users
  CORE_FEATURES: [
    ...ROLE_PERMISSIONS.ADMIN_ROLES,
    ...ROLE_PERMISSIONS.VERIFICATION_ROLES,
    ...ROLE_PERMISSIONS.SCIENTIFIC_ROLES,
    ...ROLE_PERMISSIONS.BUSINESS_ROLES,
    ...ROLE_PERMISSIONS.REGULATORY_ROLES,
    ...ROLE_PERMISSIONS.DEVELOPMENT_ROLES
  ],
  
  // ML and AI features
  ML_FEATURES: [
    ...ROLE_PERMISSIONS.ADMIN_ROLES,
    ...ROLE_PERMISSIONS.VERIFICATION_ROLES,
    ...ROLE_PERMISSIONS.SCIENTIFIC_ROLES,
    ...ROLE_PERMISSIONS.DEVELOPMENT_ROLES
  ],
  
  // XAI (Explainable AI) features - requires higher permissions
  XAI_FEATURES: [
    ...ROLE_PERMISSIONS.ADMIN_ROLES,
    ...ROLE_PERMISSIONS.VERIFICATION_ROLES,
    ...ROLE_PERMISSIONS.SCIENTIFIC_ROLES,
    ...ROLE_PERMISSIONS.REGULATORY_ROLES
  ],
  
  // Verification workflow
  VERIFICATION_WORKFLOW: [
    ...ROLE_PERMISSIONS.ADMIN_ROLES,
    ...ROLE_PERMISSIONS.VERIFICATION_ROLES,
    ...ROLE_PERMISSIONS.SCIENTIFIC_ROLES,
    ...ROLE_PERMISSIONS.DEVELOPMENT_ROLES
  ],
  
  // IoT and monitoring
  IOT_MONITORING: [
    ...ROLE_PERMISSIONS.ADMIN_ROLES,
    ...ROLE_PERMISSIONS.VERIFICATION_ROLES,
    ...ROLE_PERMISSIONS.SCIENTIFIC_ROLES,
    ...ROLE_PERMISSIONS.REGULATORY_ROLES,
    ...ROLE_PERMISSIONS.DEVELOPMENT_ROLES
  ],
  
  // Blockchain and certificates
  BLOCKCHAIN_FEATURES: [
    ...ROLE_PERMISSIONS.ADMIN_ROLES,
    ...ROLE_PERMISSIONS.VERIFICATION_ROLES,
    ...ROLE_PERMISSIONS.BUSINESS_ROLES,
    ...ROLE_PERMISSIONS.REGULATORY_ROLES,
    ...ROLE_PERMISSIONS.DEVELOPMENT_ROLES
  ],
  
  // Reporting and analytics
  REPORTING: [
    ...ROLE_PERMISSIONS.ADMIN_ROLES,
    ...ROLE_PERMISSIONS.VERIFICATION_ROLES,
    ...ROLE_PERMISSIONS.REGULATORY_ROLES
  ],
  
  // System administration
  ADMIN_FEATURES: [
    ...ROLE_PERMISSIONS.ADMIN_ROLES
  ]
};

// Helper functions for role checking
export const hasRole = (userRole, allowedRoles) => {
  if (!userRole || !allowedRoles) return false;
  return allowedRoles.includes(userRole);
};

export const hasAnyRole = (userRole, roleGroups) => {
  if (!userRole || !roleGroups) return false;
  return roleGroups.some(group => group.includes(userRole));
};

export const isAdmin = (userRole) => {
  return hasRole(userRole, ROLE_PERMISSIONS.ADMIN_ROLES);
};

export const canAccessFeature = (userRole, featureKey) => {
  const allowedRoles = FEATURE_ACCESS[featureKey];
  return hasRole(userRole, allowedRoles);
};

export const getUserRoleDisplayName = (role) => {
  // Map internal role names to user-friendly display names
  const displayNames = {
    // Admin roles
    [ROLES.ADMINISTRATOR]: 'System Administrator',
    [ROLES.ADMIN]: 'Administrator',
    
    // User roles
    [ROLES.PROJECT_DEVELOPER]: 'Project Developer',
    [ROLES.VERIFIER]: 'Carbon Verifier',
    [ROLES.SCIENTIST]: 'Environmental Scientist',
    [ROLES.RESEARCHER]: 'Climate Researcher',
    [ROLES.INVESTOR]: 'Carbon Credit Investor',
    [ROLES.BROKER]: 'Carbon Credit Broker',
    [ROLES.REGULATOR]: 'Environmental Regulator',
    [ROLES.MONITOR]: 'Environmental Monitor',
    [ROLES.AUDITOR]: 'Third-Party Auditor'
  };
  
  return displayNames[role] || role;
};

export const getRoleDescription = (role) => {
  const descriptions = {
    // Admin roles - clearly differentiated
    [ROLES.ADMINISTRATOR]: 'Complete system access and user management',
    [ROLES.ADMIN]: 'System administration with legacy access',
    
    // User roles - simplified and clear
    [ROLES.PROJECT_DEVELOPER]: 'Create and manage projects',
    [ROLES.VERIFIER]: 'Review and verify carbon claims',
    [ROLES.SCIENTIST]: 'Environmental data analysis',
    [ROLES.RESEARCHER]: 'Climate research studies',
    [ROLES.INVESTOR]: 'Investment and portfolio tracking',
    [ROLES.BROKER]: 'Carbon credit marketplace trading',
    [ROLES.REGULATOR]: 'Compliance monitoring and oversight',
    [ROLES.MONITOR]: 'Environmental data monitoring',
    [ROLES.AUDITOR]: 'Independent project auditing'
  };
  
  return descriptions[role] || 'System user';
};

// Menu item configuration using feature access
export const getMenuItemsForRole = (userRole) => {
  const menuConfig = [
    {
      text: 'Dashboard',
      path: '/dashboard',
      feature: 'CORE_FEATURES',
      description: 'Overview and quick actions'
    },
    {
      text: 'Projects',
      path: '/projects',
      feature: 'CORE_FEATURES',
      description: 'Manage carbon credit projects'
    },
    {
      text: 'Verification Hub',
      path: '/verification',
      feature: 'ADMIN_FEATURES',
      description: 'Bulk verification management (Admin only)'
    },
    {
      text: 'Explainable AI',
      path: '/xai',
      feature: 'XAI_FEATURES',
      description: 'Model explanations and transparency'
    },
    {
      text: 'IoT Sensors',
      path: '/iot',
      feature: 'IOT_MONITORING',
      description: 'Ground-based sensor data'
    },
    {
      text: 'Analytics',
      path: '/analytics',
      feature: 'CORE_FEATURES',
      description: 'Performance insights and trends'
    },
    {
      text: 'Blockchain',
      path: '/blockchain',
      feature: 'BLOCKCHAIN_FEATURES',
      description: 'Certificate verification and explorer'
    },
    {
      text: 'Reports',
      path: '/reports',
      feature: 'REPORTING',
      description: 'Verification certificates and audits'
    },
    {
      text: 'Settings',
      path: '/settings',
      feature: 'ADMIN_FEATURES',
      description: 'Account and system preferences'
    }
  ];
  
  // Filter menu items based on user role and feature access
  return menuConfig.filter(item => 
    canAccessFeature(userRole, item.feature)
  );
}; 

File: backend/utils/role_utils.py
"""
Role-based access control utilities for the Carbon Credit Verification System
Centralized role management to maintain DRY principles and consistency
"""

from typing import List, Optional
from enum import Enum

class Roles(str, Enum):
    """Enumeration of all available roles in the system"""
    # Admin roles (legacy support)
    ADMINISTRATOR = "Administrator"  # Standard admin role
    ADMIN = "admin"  # Legacy lowercase admin role
    
    # User roles
    PROJECT_DEVELOPER = "Project Developer"
    VERIFIER = "Verifier"
    SCIENTIST = "Scientist"
    RESEARCHER = "Researcher"
    INVESTOR = "Investor"
    BROKER = "Broker"
    REGULATOR = "Regulator"
    MONITOR = "Monitor"
    AUDITOR = "Auditor"

class RolePermissions:
    """Define role hierarchies and permissions"""
    # Administrative roles - highest permissions
    ADMIN_ROLES = [Roles.ADMINISTRATOR, Roles.ADMIN]
    
    # Verification and quality assurance
    VERIFICATION_ROLES = [Roles.VERIFIER, Roles.AUDITOR]
    
    # Scientific and research roles
    SCIENTIFIC_ROLES = [Roles.SCIENTIST, Roles.RESEARCHER]
    
    # Business and finance roles
    BUSINESS_ROLES = [Roles.INVESTOR, Roles.BROKER]
    
    # Regulatory and compliance roles
    REGULATORY_ROLES = [Roles.REGULATOR, Roles.MONITOR]
    
    # Development roles
    DEVELOPMENT_ROLES = [Roles.PROJECT_DEVELOPER]
    
    # All roles that can access basic features
    ALL_AUTHENTICATED_ROLES = list(Roles)

def has_role(user_role: Optional[str], allowed_roles: List[str]) -> bool:
    """
    Check if user has one of the allowed roles
    
    Args:
        user_role: The user's role
        allowed_roles: List of allowed roles
        
    Returns:
        bool: True if user has permission, False otherwise
    """
    if not user_role or not allowed_roles:
        return False
    return user_role in allowed_roles

def is_admin(user_role: Optional[str]) -> bool:
    """
    Check if user has admin privileges
    
    Args:
        user_role: The user's role
        
    Returns:
        bool: True if user is admin, False otherwise
    """
    return has_role(user_role, RolePermissions.ADMIN_ROLES)

def can_manage_projects(user_role: Optional[str]) -> bool:
    """
    Check if user can manage projects (admin or project developer)
    
    Args:
        user_role: The user's role
        
    Returns:
        bool: True if user can manage projects, False otherwise
    """
    allowed_roles = RolePermissions.ADMIN_ROLES + RolePermissions.DEVELOPMENT_ROLES
    return has_role(user_role, allowed_roles)

def can_verify_projects(user_role: Optional[str]) -> bool:
    """
    Check if user can verify projects
    
    Args:
        user_role: The user's role
        
    Returns:
        bool: True if user can verify projects, False otherwise
    """
    allowed_roles = RolePermissions.ADMIN_ROLES + RolePermissions.VERIFICATION_ROLES
    return has_role(user_role, allowed_roles)

def can_access_analytics(user_role: Optional[str]) -> bool:
    """
    Check if user can access analytics features
    
    Args:
        user_role: The user's role
        
    Returns:
        bool: True if user can access analytics, False otherwise
    """
    allowed_roles = (RolePermissions.ADMIN_ROLES + 
                    RolePermissions.VERIFICATION_ROLES + 
                    RolePermissions.SCIENTIFIC_ROLES)
    return has_role(user_role, allowed_roles)

def can_manage_users(user_role: Optional[str]) -> bool:
    """
    Check if user can manage other users
    
    Args:
        user_role: The user's role
        
    Returns:
        bool: True if user can manage users, False otherwise
    """
    return has_role(user_role, RolePermissions.ADMIN_ROLES)

def can_access_blockchain(user_role: Optional[str]) -> bool:
    """
    Check if user can access blockchain features
    
    Args:
        user_role: The user's role
        
    Returns:
        bool: True if user can access blockchain, False otherwise
    """
    allowed_roles = (RolePermissions.ADMIN_ROLES + 
                    RolePermissions.VERIFICATION_ROLES)
    return has_role(user_role, allowed_roles)

def normalize_role(role: Optional[str]) -> Optional[str]:
    """
    Normalize role string to handle case variations
    
    Args:
        role: The role string to normalize
        
    Returns:
        str: Normalized role string or None if invalid
    """
    if not role:
        return None
    
    # Handle common variations
    role_mappings = {
        'admin': Roles.ADMIN,
        'administrator': Roles.ADMINISTRATOR,
        'project developer': Roles.PROJECT_DEVELOPER,
        'verifier': Roles.VERIFIER,
        'scientist': Roles.SCIENTIST,
        'researcher': Roles.RESEARCHER,
        'investor': Roles.INVESTOR,
        'broker': Roles.BROKER,
        'regulator': Roles.REGULATOR,
        'monitor': Roles.MONITOR,
        'auditor': Roles.AUDITOR,
    }
    
    # First try exact match
    if role in [r.value for r in Roles]:
        return role
    
    # Try case-insensitive mapping
    normalized = role_mappings.get(role.lower())
    return normalized if normalized else role

# Decorator for admin-only endpoints
def require_admin(func):
    """Decorator to require admin role for endpoint access"""
    def wrapper(*args, **kwargs):
        # This would be used with dependency injection in FastAPI
        # Implementation depends on how current_user is passed
        return func(*args, **kwargs)
    return wrapper